### PR TITLE
GH-39574: [Go] Enable PollFlightInfo in Flight RPC

### LINKED ISSUE #39574: [Go] Enable using PollFlightInfo in Flight SQL servers/clients
### Describe the enhancement requested

PollFlightInfo was implemented in 14.0.0, but the wrapping we do of the gRPC client/server means more work is needed to expose it.

### Component(s)

Go

### PR DESCRIPTION
**Pull request by lidavidm**
### Rationale for this change

It's impossible to use the current bindings with PollFlightInfo. Required for apache/arrow-adbc#1457.

### What changes are included in this PR?

Add new methods that expose PollFlightInfo.

### Are these changes tested?

Yes

### Are there any user-facing changes?

Adds new methods.
* Closes: #39574

### CODE REVIEW

**Comment by lidavidm** [comment]
CC @zeroshade

File: go/arrow/flight/flightsql/client.go [hunk 1]
```diff
@@ -981,6 +1027,48 @@ func (p *PreparedStatement) Execute(ctx context.Context, opts ...grpc.CallOption
...
+			}
+			if err = wr.Close(); err != nil {
+				return nil, err
+			}
+			pstream.CloseSend()
+
+			// wait for the server to ack the result
+			if _, err = pstream.Recv(); err != nil && err != io.EOF {
+				return nil, err
+			}
```
Comment (zeroshade) [inline go/arrow/flight/flightsql/client.go:1065]: didn't we have a helper function that did this?

File: go/arrow/flight/flightsql/client.go [hunk 2]
```diff
@@ -981,6 +1027,48 @@ func (p *PreparedStatement) Execute(ctx context.Context, opts ...grpc.CallOption
...
+		}
+		return p.client.Client.PollFlightInfo(ctx, desc, opts...)
+	}
+	return p.client.Client.PollFlightInfo(ctx, retryDescriptor, opts...)
```
Comment (zeroshade) [inline go/arrow/flight/flightsql/client.go:1069]: couldn't we condense this by doing `retryDescriptor = desc` and not duplicate the line?

Comment (zeroshade) [reply go/arrow/flight/flightsql/client.go:1069]: Either that, or can we simplify and early return `if retryDescriptor != nil { return p.client.Client.PollFlightInfo.... }`

**Review by zeroshade — APPROVED** [review]
Just a few nitpicks but otherwise looks good to me

File: go/arrow/flight/flightsql/client.go (hunk 1 above)
Comment (lidavidm) [reply go/arrow/flight/flightsql/client.go:1065]: I don't think so. This is basically verbatim from the synchronous version of this method.

File: go/arrow/flight/flightsql/client.go (hunk 2 above)
Comment (lidavidm) [reply go/arrow/flight/flightsql/client.go]: Updated

**Review by zeroshade — APPROVED** [review]
LGTM

**Comment by raulcd** [comment]
@github-actions crossbow submit verify-rc-source-integration-linux-*

**Comment by raulcd** [comment]
I am verifying if this caused: https://github.com/apache/arrow/issues/41201

**Pull request merged**